quickwit-doc-mapper/src/query_builder.rs

```diff
@@ -82,41 +84,97 @@ fn has_range_clause(user_input_ast: &UserInputAst) -> bool {
 /// Tells if the query has a Term or Range node which does not
 /// specify a search field.
 fn needs_default_search_field(user_input_ast: &UserInputAst) -> bool {
+    collect_field_names(user_input_ast)
+        .iter()
+        .any(|field_opt| field_opt.is_none())
+}
+
+fn resolve_fields(schema: &Schema, field_names: &[String]) -> anyhow::Result<Vec<Field>> {
+    let mut fields = vec![];
+    for field_name in field_names {
+        let field = schema
+            .get_field(field_name)
+            .ok_or_else(|| TantivyQueryParserError::FieldDoesNotExist(field_name.clone()))?;
+        fields.push(field);
+    }
+    Ok(fields)
+}
+
+// Collects the fields names on term and range ast nodes.
+fn collect_field_names(user_input_ast: &UserInputAst) -> Vec<Option<String>> {
```

Comment:
What is the result type about? why not `Vec<String>` ?

Reply:
The prototype and the comment are however not helping.

I suggest you write the functions

```rust
fn leaves(user_input_ast: &UserInputAst) -> Vec<&UserInputLeaf>;
fn extract_field_name(leaf: &UserInputLeaf) -> Option<&str>;
// calls leaves and extract_field_name
fn needs_default_search_field(ast: &UserInputLeaf) -> bool;
// calls leaves and extract_field_name
fn field_names(ast: &UserInputAst) -> HashSet<&str>;
```

Reply:
If you think this is overkill, you can also add a comment to the function.

Reply:
I am reusing it elsewhere to test if some ast nodes don't have field name specified. 
The naming is not great, I initially called it `collect_field_nodes`